File: frontend/src/api/customers.ts
import client from './client'
import type { Customer, PaginatedResponse } from '../types'

export const getCustomers = async (params?: {
  skip?: number
  limit?: number
  search?: string
}) => {
  const { data } = await client.get<PaginatedResponse<Customer>>('/customers', { params })
  return data
}

export const getCustomer = async (id: string) => {
  const { data } = await client.get<Customer>(`/customers/${id}`)
  return data
}

export const createCustomer = async (payload: {
  full_name: string
  email?: string
  phone?: string
  company?: string
  notes?: string
}) => {
  const { data } = await client.post<Customer>('/customers', payload)
  return data
}

File: backend/app/api/v1/customers.py
import uuid
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import get_db
from app.core.middleware import get_current_user
from app.core.tenant import get_tenant_id
from app.schemas.customer import CustomerCreate, CustomerUpdate, CustomerResponse
from app.services.customer_service import CustomerService

router = APIRouter(prefix="/customers", tags=["customers"])


def get_service() -> CustomerService:
    return CustomerService(get_tenant_id())


@router.post("", response_model=CustomerResponse, status_code=201)
async def create_customer(
    data: CustomerCreate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return await get_service().create(db, data)


@router.get("", response_model=dict)
async def list_customers(
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=200),
    search: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return await get_service().get_all(db, skip, limit, search)


@router.get("/{customer_id}", response_model=CustomerResponse)
async def get_customer(
    customer_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return await get_service().get_by_id(db, customer_id)


@router.patch("/{customer_id}", response_model=CustomerResponse)
async def update_customer(
    customer_id: uuid.UUID,
    data: CustomerUpdate,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return await get_service().update(db, customer_id, data)


@router.delete("/{customer_id}")
async def delete_customer(
    customer_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user=Depends(get_current_user),
):
    return await get_service().delete(db, customer_id)